Refactor node factory API, use node factory in parser

rbuckton opened the pull request:
This overhauls our node factory API to provide the following benefits:

* Allows us to leverage the factory API from within the parser.
    * This is intended to help us ensure that the property layout of each kind of `Node` is stable between the parser, checker, and transformers to reduce polymorphism, as well as to provide a "single source of truth" with respect to how nodes are created.
* Allows us to streamline transform flag aggregation into node creation.
    * This allows us to remove transform flag logic from `binder.ts` and `transformer.ts` and ensures that transform flags will *always* be correctly aggregated during transformations.
* Extracts our auto-parenthesization logic out of the node factories.
    * This allows us to use the new node factories in the parser without the undue overhead incurred by the parenthesizer rules.
* Allows us to deprecate older node factory APIs and reduce node factory overloads.

In the future, this will also allow us to better isolate which transformer was responsible for creating nodes, as well as to possibly leverage transformers outside of the emit pipeline.

As part of these changes, all of the existing `create*` and `update*` factory functions in the `ts` namespace are being moved to a `NodeFactory` object that can be created via the internal `createNodeFactory` function. Authors of transformers can access the `NodeFactory` associated with their transformation via the `TransformationContext` provided to their transformer:

```ts
import * as ts from "typescript";
function myTransformer(context: ts.TransformationContext) {
  // old: createIdentifier("foo");

  // new (option 1):
  const { factory } = context;
  factory.createIdentifier("foo");

  // new (option 2):
  const { factory: { createIdentifier } } = context;
  createIdentifier("foo"); // same as old
}
```

In addition, the `ts.factory` object is a singleton `NodeFactory` instance, which can be used to create synthetic nodes *outside* of a transformer.

So as not to break compatibility with older API consumers, the existing `create*` and `update*` factory functions will also be available to consumers of the TypeScript language service and the TS Server API. However, these functions are considered deprecated, and will be phased out using the following (tentative) plan:

- 4.0 - Existing factory functions are marked with a `@deprecated` JSDoc comment and are no longer used by TypeScript internals.
- 4.1 (?) - The first time a deprecated factory function is called, the Language Service will attempt to write a deprecation warning message (depending on the host and whether a suitable `console.log`-like logger is available).
- 4.2 (?) - Whenever a deprecated factory function is called, an exception will be thrown.
- 4.3 (?) - Deprecated factory functions will be removed.

Remaining TODO items:
- [x] Consider renaming concise functions like `createParen` to a more expressive (and less ambiguous) name such as `createParenthesizedExpression`.
- [x] Consider making most properties of nodes `readonly`, as one goal is to treat a TypeScript AST as immutable (and should be "changed" via visitors).
- [x] Verify performance of parser use of new API.

rbuckton commented:
@typescript-bot perf test

weswigham commented on tests/baselines/reference/JSDocParsing/DocComments.parsesCorrectly.typedefTagWithChildrenTags.json:
```diff
@@ -20,30 +20,14 @@
                 "transformFlags": 0,
-                "jsDocPropertyTags": [
-                    {
+                "jsDocPropertyTags": {
```
??? Why'd this go from being serialized as an array to an object with numeric keys? Are `NodeArray`s no longer arrays?

rbuckton replied:
That's how JSON serialization handles arrays that have extra properties, so the properties aren't lost.

rbuckton commented:
@typescript-bot perf test

ajafff commented:
What's the benefit of having to use `context.factory` in transforms and how is it different from `ts.factory`?

rbuckton commented:
> What's the benefit of having to use `context.factory` in transforms and how is it different from `ts.factory`?

The current factory methods mark nodes as `Synthesized` and do some other work don't make sense for some situations. Since the new `NodeFactory` API allows for a number of additional hooks for when nodes are created and updated, it is better for a transformer to rely on the `NodeFactory` provided to it, as that will eventually allow us to be more flexible with transformations in the future.

For example, the parser now utilizes a custom `NodeFactory` that can explicitly set parent pointers as the AST is constructed:

https://github.com/microsoft/TypeScript/blob/086633f4db521efb6bc25aadaca68f457e56a95b/src/compiler/parser.ts#L695-L716

The existing factory methods will still be available for some period of time with deprecations: https://github.com/microsoft/TypeScript/blob/086633f4db521efb6bc25aadaca68f457e56a95b/src/compat/deprecations.ts. This model allows us to introduce deprecations with different levels of aggressiveness:

- Soft - Deprecations in documentation only.
- Warn - Deprecations that attempt to warn you when used (usually via `console.log` if available).
- Error - Deprecations that throw errors when used.
- Removed - The deprecated API is completely removed from the package.

rbuckton commented:
@typescript-bot perf test

rbuckton commented:
@typescript-bot perf test

amcasey commented:
@typescript-bot pack this

weswigham commented on src/compiler/debug.ts:
```diff
@@ -410,5 +470,92 @@ namespace ts {
+            const deprecation = createDeprecation(key, options);
+            const newDesc = desc.get || desc.set ? wrapAccessor(deprecation, desc) : wrapValue(deprecation, desc);
+            if (newDesc) {
+                Object.defineProperty(ns, key, newDesc);
```
This is probably going to have a bad time when refactored to modules; I think the `defineProperty` will throw when the export already exists as a non-configurable getter (ie, downlevel), and will likewise throw in native modules. It might be a good idea to preemptively change it to a usage pattern like
```ts
export const updateGetAccessor = Debug.deprecate(factory.updateGetAccessor, { since: "4.0", warnOn: "4.1", message });
```
to avoid the need for reflective reassignment. If you pull off the `message` (or the entire options object) into a local variable, it's not much more verbose than the current export+deprecate setup, but keeps the exports and deprecation for the exports on the same line (so we're not going to accidentally remove one but not the other later).

rbuckton commented:
@typescript-bot perf test

rbuckton commented:
@typescript-bot perf test

rbuckton commented:
Those are much better perf numbers. Turns out I had a bug in the caching mechanism for the various binary and unary operator factories like `createComma`, etc.

rbuckton commented:
@RyanCavanaugh - What do you think about this TODO item:
> - [ ] Consider renaming concise functions like `createParen` to a more expressive (and less ambiguous) name such as `createParenthesizedExpression`.

It would increase file size somewhat due to the less concise names, but would remove ambiguity and is clearer overall. For example, the following existing factory function names are somewhat ambiguous:

- `createParen` vs. `createParenthesizedType`. Would be clearer if renamed to `createParenthesizedExpression`.
- `createConditional` vs. `createConditionalType`. Would be clearer if renamed to `createConditionalExpression`
- `createSpread` vs. `createSpreadAssignment`. Would be clearer if renamed to `createSpreadElement`.

In general I think it would be better to adopt more descriptive names for the factory methods. You can find examples of the proposed names in the following two listings:

https://github.com/microsoft/TypeScript/blob/264b5472a077698a8ee3be8ad11d0d7a1ea18728/src/compiler/factory/nodeFactory.ts#L143-L201

https://github.com/microsoft/TypeScript/blob/264b5472a077698a8ee3be8ad11d0d7a1ea18728/src/compiler/factory/nodeFactory.ts#L225-L252

rbuckton commented:
/cc @DanielRosenwasser as he had comments on the factory method naming

rbuckton commented:
I've also created #39058 as a PR against **this** branch with the factory method renames I [proposed above](https://github.com/microsoft/TypeScript/pull/35282#issuecomment-643014388)

RyanCavanaugh commented on src/compiler/checker.ts:
```diff
@@ -14091,9 +14194,10 @@ namespace ts {
-                case SyntaxKind.Identifier:
-                case SyntaxKind.QualifiedName:
+                // Callers should first ensure this by calling `isPartOfTypeNode`
+                // TODO(rbuckton): These aren't valid TypeNodes, but we treat them as such because of `isPartOfTypeNode`, which returns `true` for things that aren't `TypeNode`s.
```
What's the follow-up needed here?

rbuckton replied:
The TODO is to eventually remove these once we fix other related code in `isPartOfTypeNode`, but that was something I considered out-of-scope for this PR. I can update the comment if necessary.

rbuckton commented:
I'm going to merge master into this one more time to make sure there aren't any new changes that need to be captured, then I can merge this to master.

nonara commented:
Nice work! Was a decision made on the renaming proposal?

rbuckton commented:
Yes, we merged https://github.com/microsoft/TypeScript/pull/39058 which used the more consistent and clearer names for factory methods.
